File: backend/app/services/scene_prompt_manager.py
"""
场景提示词管理器
"""
import json
import logging
from pathlib import Path
from typing import Dict, Any, Optional
from datetime import datetime

logger = logging.getLogger("scene_prompt_manager")


class ScenePromptManager:
    
    _base_path: Path = None
    
    @classmethod
    def _get_base_path(cls) -> Path:
        """获取基础路径"""
        if cls._base_path is None:
            cls._base_path = Path(__file__).parent.parent.parent / "config" / "prompts" / "scenes"
        return cls._base_path
    
    @classmethod
    def _ensure_dir(cls):
        """确保目录存在"""
        base_path = cls._get_base_path()
        base_path.mkdir(parents=True, exist_ok=True)
        
        # 确保模板目录存在
        templates_dir = base_path / "_templates"
        templates_dir.mkdir(parents=True, exist_ok=True)
    
    @classmethod
    def load_prompt(cls, prompt_file: str) -> Optional[str]:
        """加载提示词文件"""
        try:
            base_path = cls._get_base_path()
            file_path = base_path / prompt_file
            
            if not file_path.exists():
                logger.warning(f"Prompt file not found: {prompt_file}")
                return None
            
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            logger.debug(f"Loaded prompt: {prompt_file}")
            return content
        except Exception as e:
            logger.exception(f"Failed to load prompt {prompt_file}: {e}")
            return None
    
    @classmethod
    def save_prompt(cls, prompt_file: str, content: str) -> Dict[str, Any]:
        """保存提示词文件"""
        try:
            cls._ensure_dir()
            base_path = cls._get_base_path()
            file_path = base_path / prompt_file
            
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            
            logger.info(f"Saved prompt: {prompt_file}")
            return {"success": True, "message": "Prompt saved successfully"}
        except Exception as e:
            logger.exception(f"Failed to save prompt {prompt_file}: {e}")
            return {"success": False, "message": str(e)}
    
    @classmethod
    def delete_prompt(cls, prompt_file: str) -> Dict[str, Any]:
        """删除提示词文件"""
        try:
            base_path = cls._get_base_path()
            file_path = base_path / prompt_file
            
            if file_path.exists():
                file_path.unlink()
                logger.info(f"Deleted prompt: {prompt_file}")
            
            return {"success": True, "message": "Prompt deleted successfully"}
        except Exception as e:
            logger.exception(f"Failed to delete prompt {prompt_file}: {e}")
            return {"success": False, "message": str(e)}
    
    @classmethod
    def list_prompts(cls) -> Dict[str, Any]:
        """列出所有提示词文件"""
        try:
            base_path = cls._get_base_path()
            
            if not base_path.exists():
                return {"success": True, "data": []}
            
            prompts = []
            for file_path in base_path.glob("*.txt"):
                if file_path.is_file():
                    stat = file_path.stat()
                    prompts.append({
                        "fileName": file_path.name,
                        "fileSize": stat.st_size,
                        "lastModified": datetime.fromtimestamp(stat.st_mtime).isoformat()
                    })
            
            # 按名称排序
            prompts.sort(key=lambda x: x["fileName"])
            
            return {"success": True, "data": prompts}
        except Exception as e:
            logger.exception(f"Failed to list prompts: {e}")
            return {"success": False, "message": str(e)}
    
    @classmethod
    def load_template(cls, template_type: str) -> Optional[str]:
        """加载模板"""
        try:
            template_file = f"{template_type}.txt"
            base_path = cls._get_base_path()
            file_path = base_path / "_templates" / template_file
            
            if not file_path.exists():
                logger.warning(f"Template not found: {template_file}")
                return None
            
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            return content
        except Exception as e:
            logger.exception(f"Failed to load template {template_type}: {e}")
            return None
    
    @classmethod
    def build_prompt(cls, prompt_template: str, context: Dict[str, Any]) -> str:
        """构建提示词，替换变量"""
        try:
            result = prompt_template
            
            # 替换变量
            variables = {
                "{scene_code}": str(context.get("scene_code", "")),
                "{scene_name}": str(context.get("scene_name", "")),
                "{form_code}": str(context.get("form_code", "")),
                "{available_tools}": ", ".join(context.get("available_tools", [])),
                "{user_input}": str(context.get("user_input", "")),
                "{current_date}": datetime.now().strftime("%Y-%m-%d"),
                "{current_time}": datetime.now().strftime("%H:%M:%S"),
            }
            
            for var, value in variables.items():
                result = result.replace(var, value)
            
            # 构建工具信息
            if "tools_info" in context:
                tools_info = context["tools_info"]
                if isinstance(tools_info, str):
                    result = result.replace("{tools_info}", tools_info)
                elif isinstance(tools_info, list):
                    tools_str = "\n".join([f"- {t}" for t in tools_info])
                    result = result.replace("{tools_info}", tools_str)
                else:
                    result = result.replace("{tools_info}", "")
            
            return result
        except Exception as e:
            logger.exception(f"Failed to build prompt: {e}")
            return prompt_template
    
    @classmethod
    def create_prompt_from_template(cls, template_type: str, scene_info: Dict[str, Any]) -> Dict[str, Any]:
        """从模板创建提示词"""
        try:
            template = cls.load_template(template_type)
            if not template:
                return {"success": False, "message": f"Template {template_type} not found"}
            
            prompt_content = cls.build_prompt(template, scene_info)
            
            return {"success": True, "data": prompt_content}
        except Exception as e:
            logger.exception(f"Failed to create prompt from template: {e}")
            return {"success": False, "message": str(e)}

File: backend-app/src/main/java/com/sitech/prodai/service/ScenePromptManager.java
package com.sitech.prodai.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 场景提示词文件管理器 —— 对齐 Python {@code app/services/scene_prompt_manager.py::ScenePromptManager}。
 *
 * <p>Python 版本从 {@code backend/config/prompts/scenes/} 读写提示词文件。
 * Java 版本双轨制：
 * <ul>
 *   <li>读取：优先从 classpath:prompts/scenes/ 读取（对齐 Python 文件数据源）</li>
 *   <li>写入：写入工作目录下的 {@code prompts/scenes/}（Spring Boot 不能写入 jar 内部，
 *       故使用外部目录保存运行时生成的提示词）</li>
 * </ul>
 *
 * <p>外部目录可通过配置项 {@code app.prompt-scenes-dir} 自定义，默认 {@code ./prompts/scenes}。
 */
@Service
public class ScenePromptManager {

    private static final Logger log = LoggerFactory.getLogger(ScenePromptManager.class);
    private static final DateTimeFormatter TIME_FMT = DateTimeFormatter.ofPattern("HH:mm:ss");

    private final PathMatchingResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();
    private final Path externalBasePath;
    private final Path externalTemplatesDir;

    public ScenePromptManager(@Value("${app.prompt-scenes-dir:prompts/scenes}") String externalDir) {
        this.externalBasePath = Paths.get(externalDir).toAbsolutePath().normalize();
        this.externalTemplatesDir = externalBasePath.resolve("_templates");
        try {
            Files.createDirectories(externalBasePath);
            Files.createDirectories(externalTemplatesDir);
        } catch (IOException e) {
            log.warn("[ScenePromptManager] 初始化目录失败: {}", e.getMessage());
        }
    }

    /** 对齐 Python load_prompt —— 优先外部目录，回退 classpath */
    public String loadPrompt(String promptFile) {
        // 1. 先查外部目录
        Path external = externalBasePath.resolve(promptFile);
        if (Files.exists(external)) {
            try {
                return Files.readString(external, StandardCharsets.UTF_8);
            } catch (IOException e) {
                log.warn("[ScenePromptManager] 读取外部提示词失败 {}: {}", promptFile, e.getMessage());
            }
        }
        // 2. 回退 classpath
        try {
            Resource resource = resolver.getResource("classpath:prompts/scenes/" + promptFile);
            if (resource.exists()) {
                try (var in = resource.getInputStream()) {
                    return new String(in.readAllBytes(), StandardCharsets.UTF_8);
                }
            }
        } catch (IOException e) {
            log.warn("[ScenePromptManager] 读取 classpath 提示词失败 {}: {}", promptFile, e.getMessage());
        }
        return null;
    }

    /** 对齐 Python save_prompt —— 写入外部目录 */
    public Map<String, Object> savePrompt(String promptFile, String content) {
        try {
            Files.createDirectories(externalBasePath);
            Path target = externalBasePath.resolve(promptFile);
            Files.writeString(target, content == null ? "" : content, StandardCharsets.UTF_8);
            log.info("[ScenePromptManager] 保存提示词: {}", promptFile);
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", true);
            body.put("message", "Prompt saved successfully");
            return body;
        } catch (IOException e) {
            log.error("[ScenePromptManager] 保存提示词失败 {}", promptFile, e);
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", false);
            body.put("message", str(e));
            return body;
        }
    }

    /** 对齐 Python delete_prompt —— 从外部目录删除 */
    public Map<String, Object> deletePrompt(String promptFile) {
        try {
            Path target = externalBasePath.resolve(promptFile);
            if (Files.exists(target)) {
                Files.delete(target);
                log.info("[ScenePromptManager] 删除提示词: {}", promptFile);
            }
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", true);
            body.put("message", "Prompt deleted successfully");
            return body;
        } catch (IOException e) {
            log.error("[ScenePromptManager] 删除提示词失败 {}", promptFile, e);
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", false);
            body.put("message", str(e));
            return body;
        }
    }

    /** 对齐 Python list_prompts —— 合并外部目录与 classpath 中的 .txt 文件 */
    public Map<String, Object> listPrompts() {
        List<String> names = new ArrayList<>();
        // 1. 外部目录
        try (var stream = Files.list(externalBasePath)) {
            stream.filter(Files::isRegularFile)
                    .filter(p -> p.getFileName().toString().endsWith(".txt"))
                    .filter(p -> !p.getFileName().toString().startsWith("_"))
                    .forEach(p -> names.add(p.getFileName().toString()));
        } catch (IOException e) {
            log.warn("[ScenePromptManager] 列出外部目录失败: {}", e.getMessage());
        }
        // 2. classpath
        try {
            Resource[] resources = resolver.getResources("classpath*:prompts/scenes/*.txt");
            for (Resource r : resources) {
                String name = r.getFilename();
                if (name != null && name.endsWith(".txt") && !name.startsWith("_") && !names.contains(name)) {
                    names.add(name);
                }
            }
        } catch (IOException e) {
            log.warn("[ScenePromptManager] 列出 classpath 提示词失败: {}", e.getMessage());
        }
        names.sort(Comparator.naturalOrder());
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("success", true);
        body.put("data", names);
        return body;
    }

    /** 对齐 Python load_template —— 优先外部 _templates 目录，回退 classpath */
    public String loadTemplate(String templateType) {
        String filename = templateType + ".txt";
        // 1. 外部目录
        Path external = externalTemplatesDir.resolve(filename);
        if (Files.exists(external)) {
            try {
                return Files.readString(external, StandardCharsets.UTF_8);
            } catch (IOException e) {
                log.warn("[ScenePromptManager] 读取外部模板失败 {}: {}", templateType, e.getMessage());
            }
        }
        // 2. classpath
        try {
            Resource resource = resolver.getResource("classpath:prompts/scenes/_templates/" + filename);
            if (resource.exists()) {
                try (var in = resource.getInputStream()) {
                    return new String(in.readAllBytes(), StandardCharsets.UTF_8);
                }
            }
        } catch (IOException e) {
            log.warn("[ScenePromptManager] 读取 classpath 模板失败 {}: {}", templateType, e.getMessage());
        }
        return null;
    }

    /** 对齐 Python build_prompt —— 替换模板变量 */
    public String buildPrompt(String promptTemplate, Map<String, Object> context) {
        if (promptTemplate == null) {
            return "";
        }
        String result = promptTemplate;
        result = result.replace("{scene_code}", str(context.get("scene_code")));
        result = result.replace("{scene_name}", str(context.get("scene_name")));
        result = result.replace("{form_code}", str(context.get("form_code")));
        result = result.replace("{user_input}", str(context.get("user_input")));
        result = result.replace("{current_date}", LocalDateTime.now().toLocalDate().toString());
        result = result.replace("{current_time}", LocalDateTime.now().toLocalTime().format(TIME_FMT));

        // available_tools / tools_info
        Object availableTools = context.get("available_tools");
        if (availableTools instanceof List<?> list) {
            result = result.replace("{available_tools}", String.join(", ", list.stream()
                    .map(String::valueOf)
                    .toList()));
        } else {
            result = result.replace("{available_tools}", str(availableTools));
        }

        Object toolsInfo = context.get("tools_info");
        if (toolsInfo instanceof String s) {
            result = result.replace("{tools_info}", s);
        } else if (toolsInfo instanceof List<?> list) {
            StringBuilder sb = new StringBuilder();
            for (Object t : list) {
                sb.append("- ").append(t).append("\n");
            }
            result = result.replace("{tools_info}", sb.toString());
        } else if (toolsInfo != null) {
            result = result.replace("{tools_info}", String.valueOf(toolsInfo));
        }
        return result;
    }

    /** 对齐 Python create_prompt_from_template */
    public Map<String, Object> createPromptFromTemplate(String templateType, Map<String, Object> sceneInfo) {
        try {
            String template = loadTemplate(templateType);
            if (template == null) {
                Map<String, Object> body = new LinkedHashMap<>();
                body.put("success", false);
                body.put("message", "Template " + templateType + " not found");
                return body;
            }
            String promptContent = buildPrompt(template, sceneInfo);
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", true);
            body.put("data", promptContent);
            return body;
        } catch (Exception e) {
            log.error("[ScenePromptManager] 从模板创建提示词失败 {}", templateType, e);
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("success", false);
            body.put("message", str(e));
            return body;
        }
    }

    /** 获取外部基础目录，供调试或外部 API 使用 */
    public Path getExternalBasePath() {
        return externalBasePath;
    }

    private String str(Object value) {
        return value == null ? "" : String.valueOf(value);
    }

    private String str(Throwable e) {
        return e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage();
    }
}
